bulk-model-sync-lib/src/commonMain/kotlin/org/modelix/model/sync/bulk/ModelImporter.kt

```diff
@@ -121,52 +122,127 @@ class ModelImporter(private val root: INode, private val continueOnError: Boolea
     private fun syncChildren(node: INode, data: NodeData, progressReporter: ProgressReporter) {
         val allRoles = (data.children.map { it.role } + node.allChildren.map { it.roleInParent }).distinct()
         for (role in allRoles) {
-            val expectedNodes = data.children.filter { it.role == role }
-            val existingNodes = node.getChildren(role).toList()
+            syncChildRole(data, role, node, progressReporter)
+        }
+    }
 
-            // optimization that uses the bulk operation .addNewChildren
-            if (existingNodes.isEmpty() && expectedNodes.all { originalIdToExisting[it.originalId()] == null }) {
-                node.addNewChildren(role, -1, expectedNodes.map { it.concept?.let { ConceptReference(it) } }).zip(expectedNodes).forEach { (newChild, expected) ->
+    private fun syncChildRole(
+        data: NodeData,
+        role: String?,
+        node: INode,
+        progressReporter: ProgressReporter,
+    ) {
+        val expectedNodes = data.children.filter { it.role == role }
```

Comment:
The code here is somewhat performance critical. Now that I see this (already existing) code fragment: The list of children is iterated once to compute all roles and then again for each role. I suspect there is a more efficient version that builds up a map first to avoid the repeated full iteration through all children.

Reply:
For now this is out of scope.
The needed change is more complicated, because `expectedParent.children` changes might change in between iterations.